### PR TITLE
[mono][interp] Ensure call offset is greater than offset of other active calls

### LINKED ISSUE #80175: [mono][interp] Offset allocation of call args is incorrect
The algorithm for allocating offsets for the arguments of a call must respect the following constraint: The base offset of the call must be greater than the offset of any argument of other active call args. (An active call is a call that has not yet been made and at least one of its arguments has been set already). This constraint must be respected because a call can freely overwrite the stack starting at the base offset.

The current algorithm doesn't fully respect this constraint but the issue does not reproduce currently by chance. Consider the following test case (built in Release).

```
using System;

public class Program
{
	public static int global1 = 1;

	public static void Call1 (int arg)
	{
		Console.WriteLine (arg);
		arg = 0;
	}

	public static void Main (string[] args)
	{
		int call2_arg1 = global1;
		Call1 (global1); // 1st
		int call3_arg1 = global1;
		Call1 (call2_arg1); // 2nd
		Call1 (call3_arg1); // 3rd
		// Should print 1, 1, 1. If buggy, call2_arg1 gets overwritten with 0 during 1st call
	}
}
```

The current algorithm works as follows. 2nd call becomes first active call because `call2_arg` gets stored first. 1st call becomes resolved with only one argument. Because 2nd call is active, the argument of 1st call will have to get stored at an offset greater than 2nd call param_area. Because the param area size of 2nd call is 8, the base offset of 1st call is determined to be 8, which is grossly incorrect, because the base offset of 2nd call hasn't yet been determined, while the current offset is basically assuming that it would be 0. In test case, after 1st call, 3rd call becomes active, which forces 2nd call to have its offset higher than 3rd call, therefore not 0.

As a fix, 1st call offset computation should somehow be deferred once 2nd call offset has been computed.

This testcase can be made to fail by simply removing the `cprop dreg` optimization

### PR DESCRIPTION
**Pull request by kotlarmilos**
This PR fixes https://github.com/dotnet/runtime/issues/80175 by introducing deferred calls resolution. The algorithm for allocating offsets for the arguments of a call must ensure that the base offset of the call must be greater than the offset of any argument of other active call args.

For each call, function `end_active_call` computes the call offset of each call argument starting from a base offset. The base offset is computed as max offset of all call offsets on which the call depends. Deferred call stack ensures that all call offsets on which the call depends are calculated before the call resolution.

### CODE REVIEW

File: src/mono/mono/mini/interp/transform.c [hunk 1]
```diff
@@ -3655,10 +3661,10 @@ interp_transform_call (TransformData *td, MonoMethod *method, MonoMethod *target
 			}
 #endif
 		}
-		td->last_ins->flags |= INTERP_INST_FLAG_CALL;
 	}
 	td->ip += 5;
-	td->last_ins->info.call_args = call_args;
+	init_last_ins_call (td);
+	td->last_ins->info.call_info->call_args = call_args;
```
Comment (BrzVlad) [inline src/mono/mono/mini/interp/transform.c:3667]: The `call_args` field was incorrectly set here. We should only init in places where `INTERP_INST_FLAG_CALL` is set, aka above

File: src/mono/mono/mini/interp/transform.h [hunk 2]
```diff
@@ -82,11 +83,9 @@ struct _InterpInst {
 	union {
 		InterpBasicBlock *target_bb;
 		InterpBasicBlock **target_bb_table;
-		// For call instructions, this represents an array of all call arg vars
-		// in the order they are pushed to the stack. This makes it easy to find
-		// all source vars for these types of opcodes. This is terminated with -1.
-		int *call_args;
+		InterpCallInfo *call_info;
 	} info;
+	InterpCallInfo *call_info;
```
Comment (BrzVlad) [inline src/mono/mono/mini/interp/transform.h:88]: duplicate addition

File: src/mono/mono/mini/interp/transform.c [hunk 3]
```diff
@@ -10054,61 +10038,92 @@ add_active_call (TransformData *td, ActiveCalls *ac, InterpInst *call)
...
-
-		call_args = (int*)mono_mempool_alloc (td->mempool, 3 * sizeof (int));
-		call_args [0] = new_var;
-		call_args [1] = -1;
+	// Push active call that should be resolved onto the stack
+	call->call_info = (InterpCallInfo*)mono_mempool_alloc (td->mempool, sizeof (InterpCallInfo));
+	call->call_info->call_deps = NULL;
+	for (int i = 0; i < ac->active_calls_count; i++)
+		call->call_info->call_deps = g_slist_prepend_mempool (td->mempool, call->call_info->call_deps, ac->active_calls [i]);
+	ac->deferred_calls = g_slist_prepend_mempool (td->mempool, ac->deferred_calls, call);
```
Comment (BrzVlad) [inline src/mono/mono/mini/interp/transform.c:10088]: We seem to be allocating a list entry for the most common case of no active calls / resolve arg offsets immediately.

**Comment by BrzVlad** [comment]
Once all issues are fixed I would be interested in some quick numbers for compilation time change, just in case. You could run `System.Runtime.Tests` with tiering disabled, and use `MONO_TIME_TRACK` on `interp_alloc_offsets`, with and without your change.

File: src/mono/mono/mini/interp/transform.c (hunk 1 above)
Comment (kotlarmilos) [reply src/mono/mono/mini/interp/transform.c:3667]: In `interp_transform_call` there is https://github.com/dotnet/runtime/blob/main/src/mono/mono/mini/interp/transform.c#L3556-L3577 that doesn't set `INTERP_INST_FLAG_CALL` which is probably correct. I will add there allocation of `InterpCallInfo` structure.

File: src/mono/mono/mini/interp/transform.c [hunk 4]
```diff
@@ -3655,10 +3662,12 @@ interp_transform_call (TransformData *td, MonoMethod *method, MonoMethod *target
 			}
 #endif
 		}
-		td->last_ins->flags |= INTERP_INST_FLAG_CALL;
+		init_last_ins_call (td);
```
Comment (BrzVlad) [inline src/mono/mono/mini/interp/transform.c:3665]: we should set `call_args` here and not allocate the call info below, for non-call instructions

File: src/mono/mono/mini/interp/transform.c [hunk 5]
```diff
@@ -10054,61 +10041,97 @@ add_active_call (TransformData *td, ActiveCalls *ac, InterpInst *call)
...
-		ac->active_calls = (ActiveCall*)mono_mempool_alloc (td->mempool, ac->active_calls_capacity * sizeof (ActiveCall));
-		memcpy (ac->active_calls, old, ac->active_calls_count * sizeof (ActiveCall));
+		ac->active_calls = (InterpInst**)mono_mempool_alloc (td->mempool, ac->active_calls_capacity * sizeof (InterpInst*));
+		memcpy (ac->active_calls, old, ac->active_calls_count * sizeof (InterpInst*));
 	}
-
-	ac->active_calls [ac->active_calls_count].call = call;
-	ac->active_calls [ac->active_calls_count].param_size = get_call_param_size (td, call);
-	ac->param_size += ac->active_calls [ac->active_calls_count].param_size;
+	ac->active_calls [ac->active_calls_count] = (InterpInst*)mono_mempool_alloc (td->mempool, sizeof (InterpInst));
```
Comment (BrzVlad) [inline src/mono/mono/mini/interp/transform.c:10049]: this code seems accidental

File: src/mono/mono/mini/interp/transform.c (hunk 4 above)
Comment (kotlarmilos) [reply src/mono/mono/mini/interp/transform.c:3672]: Now, there is a check `if (td->last_ins->flags & INTERP_INST_FLAG_CALL) {` so I haven't set `call_args` here. Let me know if it should be updated.

**Comment by kotlarmilos** [comment]
@BrzVlad good catch in https://github.com/dotnet/runtime/pull/80795, it fixed the failing tests.

**Comment by kotlarmilos** [comment]
> Once all issues are fixed I would be interested in some quick numbers for compilation time change, just in case. You could run `System.Runtime.Tests` with tiering disabled, and use `MONO_TIME_TRACK` on `interp_alloc_offsets`, with and without your change.

Compilation time of `interp_alloc_offsets` for System.Runtime.Tests with tiering disabled.

Test | [Incorrect offset allocation fix branch](https://github.com/dotnet/runtime/commit/ad52c218ee2835c983d0e52fc5e4f38643254325) | [Main branch](https://github.com/dotnet/runtime/commit/f429780c9ccce3546e9c9e25c05ed083318428bd)
--|--|--
System.Runtime.Tests | 24μs | 23μs

File: src/mono/mono/mini/interp/transform.c [hunk 6]
```diff
@@ -10188,61 +10171,96 @@ add_active_call (TransformData *td, ActiveCalls *ac, InterpInst *call)
...
-		alloc_var_offset (td, new_var, &start_offset);
-
-		call_args = (int*)mono_mempool_alloc (td->mempool, 3 * sizeof (int));
-		call_args [0] = new_var;
-		call_args [1] = -1;
+		// If no other active calls, current active call and all deferred calls can be resolved from the stack
+		InterpInst *deferred_call = call;
+		while (deferred_call) {
+			// `base_offset` is a relative offset (to the start of the call args stack) where the args for this call reside.
+			int base_offset = 0;
```
Comment (BrzVlad) [inline src/mono/mono/mini/interp/transform.c:10226]: I think a comment here explaining why `call_offset` is guaranteed to be set for all deps wouldn't hurt. Something like: `The deps for a call represent the list of active calls at the moment when the call ends. This means that all deps for a call end after the call in question. Given we iterate over the list of deferred calls from the last to the first one to end, all deps of a call are guaranteed to have been processed at this point.`

File: src/mono/mono/mini/interp/transform.c [hunk 7]
```diff
@@ -10188,61 +10171,96 @@ add_active_call (TransformData *td, ActiveCalls *ac, InterpInst *call)
...
 }
 
+/**
+ * Function allocates offsets of resolved calls following a constraint
+ * where the base offset of a call must be greater than the offset of any argument of other active call args.
+ *
+ * Function first removes the call from an array of active calls. If a match is found,
+ * the call is removed from the array by moving the last entry into its place. Otherwise, it is a call without arguments.
+ *
+ * Call is push onto the stack with an array of other active calls that need to be resolved first.
```
Comment (BrzVlad) [inline src/mono/mono/mini/interp/transform.c:10193]: I think this comment is outdated. You might want to double check the explanation for this method.

File: src/mono/mono/mini/interp/transform.c [hunk 8]
```diff
@@ -1218,6 +1218,13 @@ store_local (TransformData *td, int local)
 		td->last_ins->data [0] = GINT32_TO_UINT16 (td->locals [local].size);
 }
 
+static void
+init_last_ins_call (TransformData *td) {
```
Comment (BrzVlad) [inline src/mono/mono/mini/interp/transform.c:1222]: newline here

File: src/mono/mono/mini/interp/transform.c (hunk 7 above)
Comment (kotlarmilos) [reply src/mono/mono/mini/interp/transform.c]: Updated.

File: src/mono/mono/mini/interp/transform.c (hunk 6 above)
Comment (kotlarmilos) [reply src/mono/mono/mini/interp/transform.c:10228]: Added.

File: src/mono/mono/mini/interp/transform.c [hunk 9]
```diff
@@ -10188,61 +10172,97 @@ add_active_call (TransformData *td, ActiveCalls *ac, InterpInst *call)
...
 }
 
+/**
+ * Function allocates offsets of resolved calls following a constraint
+ * where the base offset of a call must be greater than the offset of any argument of other active call args.
+ *
+ * Function first removes the call from an array of active calls. If a match is found,
+ * the call is removed from the array by moving the last entry into its place. Otherwise, it is a call without arguments.
+ *
+ * If there are active calls, the call is push onto the stack with an array of other active calls on which the call in question depends.
```
Comment (BrzVlad) [inline src/mono/mono/mini/interp/transform.c:10194]: I was referring to 

> call is push onto the stack with an array of other active calls

Seemed like old comment from your original implementation where we were pushing the current call together with all active calls. Which we no longer do now.

Comment (kotlarmilos) [reply src/mono/mono/mini/interp/transform.c:10194]: If there are active calls, it is pushed onto the stack as a deferred call. It also contains an array of other active calls. Let me change it to be clear that the array is not pushed onto the stack but exists as a part of InterpInst.

**Review by BrzVlad — APPROVED** [review]

**Pull request merged**